openvdb/points/AttributeArray.h

```diff
@@ -1266,9 +1274,14 @@ TypedAttributeArray<ValueType_, Codec_>::unregisterType()
 
 template<typename ValueType_, typename Codec_>
 inline typename TypedAttributeArray<ValueType_, Codec_>::Ptr
-TypedAttributeArray<ValueType_, Codec_>::create(Index n, Index stride, bool constantStride)
+TypedAttributeArray<ValueType_, Codec_>::create(Index n, Index stride, bool constantStride,
+    const Metadata* metadata)
 {
-    return Ptr(new TypedAttributeArray(n, stride, constantStride));
+    const TypedMetadata<ValueType>* typedMetadata = metadata ?
+        dynamic_cast<const TypedMetadata<ValueType>*>(metadata) : nullptr;
+
+    return Ptr(new TypedAttributeArray(n, stride, constantStride,
+        typedMetadata ? typedMetadata->value() : zeroVal<ValueType>()));
```

Comment:
Perhaps two ::create methods on the derive types should exist so you can avoid this dynamic_cast. For example, `appendAttribute` already knows the type of the attribute, so the metadata shouldn't be dynamically cast for every leaf. Any signature which is templated on the attr type should enforce it on the metadata as well, i.e:

```
template <typename ValueType,
          typename CodecType = NullCodec,
          typename PointDataTreeT = PointDataTree>
inline void appendAttribute(PointDataTreeT& tree,
                            const std::string& name,
                            const ValueType& uniformValue =
                                point_attribute_internal::Default<ValueType>::value(),
                            const Index strideOrTotalSize = 1,
                            const bool constantStride = true,
                            const TypedMetadata<ValueType>* defaultValue = nullptr,
                            const bool hidden = false,
                            const bool transient = false);
```

Reply:
I guess through the current factory methods this isn't possible more the higher level API functions which don't runtime branch. Really what would be great is templated methods at every stage of the hierarchy. i.e. appendAttribute<ValueType> on the AttributeSet which can call the derived AttributeArray constructor

Reply:
I'll make the change to enforce the derived Metadata type in this value-typed `appendAttribute` (this is only currently used in `PointUtils.cc`). However, the ripple-out effect of enforcing this at all levels just to avoid the dynamic_cast is quite large (and potentially not even possible without some runtime branching), so I don't think it's worth addressing that at this stage.

Reply:
You shouldn't need any runtime branching through the templated methods. You only need templated method on AttributeSet::appendAttribute and potentially on the LeafNode::appendAttribute too, as well as a new constructor on the TypedAttributeArray. The problem currently is that the templated methods generate the value type strings and then call the non-templated methods - which makes them a bit redundant (performance wise) bar their ease of use. They shouldn't ever need to go to their "registered" string type and instead resolve all calls at compile time. Were it the other way around (with the bulk of the logic living in the templated methods and providing the base lib support via runtime branching for the non-templated ones) you'd get the optimal behaviour at the cost of limiting support through the NamePair API.

Reply:
I'm not suggesting this be done now btw

Reply:
In theory yes, but the types are still stored as strings in the `Descriptor `which you'll almost always want to be modifying too. Aside from the dynamic_cast, the only thing you would really save is the map lookup of the functor that creates the `AttributeArray `(https://github.com/AcademySoftwareFoundation/openvdb/blob/master/openvdb/points/AttributeArray.cc#L104). By far the bigger performance issue was the locking of the mutex in accessing this map (which we've now resolved), so I'd say it's a pretty minor issue to address.

I believe what you're suggesting would need to be an ABI=8 anyway as it requires adding a new constructor to the `TypedAttributeArray`.

(As an aside, I think we could probably get a bigger performance win by making the `std::map` of the AttributeFactoryMap into a `std::unordered_map`.)

Reply:
@Idclip - Addressed your feedback regarding the function signature of appendAttribute() including the derived metadata type.